src/pint/models/binary_piecewise.py

```diff
@@ -0,0 +1,465 @@
+            if param is not None:
+                # if parameter value or unit unset, set with default according to param
+                if param_unit is None:
+                    param_unit = (getattr(self, param)).units
+                if paramx is None:
+                    paramx = (getattr(self, param)).value
+        # check if name exits and is currently available
+        if param is None or param not in ["T0", "A1"]:
+            raise AttributeError(
+                "param '%s' not found. Please choose another. Currently implemented: 'T0' or 'A1' "
+                % param
+            )
+
+        if param == "A1":
+            self.add_param(
+                prefixParameter(
+                    name=param + f"X_{i}",
+                    units=param_unit,
+                    value=paramx,
+                    unit_template=lambda x: param_unit,
+                    description="Parameter" + param + "variation",
+                    description_template=lambda x: param,
+                    parameter_type="float",
+                    frozen=False,
+                )
+            )
+        elif param == "T0":
+            self.add_param(
+                prefixParameter(
+                    name=param + f"X_{i}",
+                    units=param_unit,
+                    value=paramx,
+                    unit_template=lambda x: param_unit,
+                    description="Parameter" + param + "variation",
+                    description_template=lambda x: param,
+                    parameter_type="float",
+                    frozen=False,
+                )
+            )
+
+    def lock_groups(self):
+        self.validate()
+        self.update_binary_object(None)
+        self.setup()
+
+    def setup(self):
+        """Raises
+        ------
+        ValueError
+                if there are values that have been added without name/ranges associated (should only be raised if add_piecewise_param has been side-stepped with an alternate method)
+        """
+        super().setup()
+        for bpar in self.params:
+            self.register_deriv_funcs(self.d_binary_delay_d_xxxx, bpar)
+        # Setup the model isinstance
+        self.binary_instance = self.binary_model_class()
+        # piecewise T0's
+        T0X_mapping = self.get_prefix_mapping_component("T0X_")
+        T0Xs = {}
+        # piecewise A1's (doing piecewise A1's requires more thought and work)
+        A1X_mapping = self.get_prefix_mapping_component("A1X_")
+        A1Xs = {}
+        # piecewise parameter ranges PLB-piece lower bound
+        XR1_mapping = self.get_prefix_mapping_component("PLB_")
+        XR1s = {}
+        # piecewise parameter ranges PUB-piece upper bound
+        XR2_mapping = self.get_prefix_mapping_component("PUB_")
+        XR2s = {}
+
+        for t0n in T0X_mapping.values():
+            T0Xs[t0n] = getattr(self, t0n).quantity
+
+        for t0_name, t0_value in T0Xs.items():
```

Comment:
would it make more sense to do a single loop over the PLB/PUB mappings?  And then for each index check whether there was a A1X or T1X value and set them if so?

Reply:
Implemented the sorting algorithm described